### PR TITLE
Auto Query Fixer: implement the FixDuplicateColumns and FixColumnNotGrouped methods

### PR DESCRIPTION
**Pull request by mingen-pan**
implement the FixDuplicateColumns and FixColumnNotGrouped methods in ZetaSQL Helper.
Test cases are included.
Readme and Dockerfile are added.

### CODE REVIEW

File: tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+#include "absl/strings/str_cat.h"
+#include "zetasql/parser/parser.h"
+#include "zetasql_helper/util/util.h"
+
+
+namespace bigquery::utils::zetasql_helper {
+
+
+absl::Status
+FixColumnNotGrouped(absl::string_view query, absl::string_view missing_column, int line_number, int column_number,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc:28]: Can you use absl::StatusOr<string> if it is released?  so you don't have pass output as string*?

File: tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  missing_column = remove_backtick(missing_column);
+
+  auto offset = get_offset(query, line_number, column_number);
+  if (offset == -1) {
+    return absl::Status(absl::StatusCode::kInvalidArgument, "Line and/or column numbers are incorrect");
+  }
+  auto select_node = find_select_node_having_column(parser_output->statement(), offset, missing_column);
+  if (select_node == nullptr) {
+    return absl::Status(absl::StatusCode::kInvalidArgument, "Cannot locate the ungrouped column");
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc:46]: nit: Cannot locate the ungrouped column "."

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// duplicate_column_name: name of the duplicate columns
+//
+// The function will assign the fixed query to the fixed_query pointer
+absl::Status FixDuplicateColumns(absl::string_view query, absl::string_view duplicate_column_name,
+                                 std::string* fixed_query);
+
+
+// Find the ASTSelectList Node having duplicate columns, whose name is the input name.
+const zetasql::ASTSelectList*
+find_select_list_with_duplicate_columns(const zetasql::ASTNode& node, absl::string_view column_name);
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h:39]: c++ style guide recommends function naming as "AddTableEntry"...

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+// Get the column name of an ASTSelectColumn node.
+std::string get_column_name(const zetasql::ASTSelectColumn* column_node);
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h:42]: GetColumnName(

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+// Change the duplicate names to their new names by adding numeric suffix after them.
+void
+replace_duplicate_columns_of_select_list(const zetasql::ASTSelectList& select_list, absl::string_view column_name,
+                                         zetasql_base::UnsafeArena* arena, zetasql::IdStringPool* id_string_pool);
+
+// Update or create an alias node of an ASTSelectColumn
+void update_alias(zetasql::ASTSelectColumn* column_node, absl::string_view new_alias,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h:50]: UpdateAlias

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                  zetasql_base::UnsafeArena* arena, zetasql::IdStringPool* id_string_pool);
+
+// Create an alias node of an ASTSelectColumn
+zetasql::ASTAlias* create_alias_node(absl::string_view new_alias, zetasql_base::UnsafeArena* arena,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h:54]: same here.

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc [hunk 7]
```diff
@@ -81,5 +83,35 @@ absl::Status ZetaSqlHelperLocalServiceImpl::GetAllKeywords(const GetAllKeywordsR
...
+    FixColumnNotGroupedResponse* response) {
+
+  std::string fix_query;
+  ZETASQL_RETURN_IF_ERROR(
+      ::bigquery::utils::zetasql_helper::FixColumnNotGrouped(
+          request.query(), request.missing_column(), request.line_number(), request.column_number(),
+          &fix_query
+      ));
+
+  response->set_fixed_query(std::string(fix_query));
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:98]: why need cast here? it is std::string right ?

File: tools/zetasql_helper/zetasql_helper/local_service/local_service_test.cc [hunk 8]
```diff
@@ -89,12 +89,57 @@ TEST_F(LocalServiceTest, LocateTableRangesTest) {
 }
 
 
-TEST_F(LocalServiceTest, GetAllKeywords) {
+TEST_F(LocalServiceTest, GetAllKeywordsTest) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service_test.cc:92]: nit: stylewise you don't need the GetAllKeywords"Test"
just GetAllKeywords

File: tools/zetasql_helper/zetasql_helper/local_service/local_service_test.cc [hunk 9]
```diff
@@ -89,12 +89,57 @@ TEST_F(LocalServiceTest, LocateTableRangesTest) {
...
   GetAllKeywordsRequest request;
   GetAllKeywordsResponse response;
   GetService().GetAllKeywords(nullptr, &request, &response);
 
   EXPECT_EQ(231, response.keywords().size());
 }
+
+TEST_F(LocalServiceTest, FixDuplicateColumnsTest) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service_test.cc:100]: nit: FixDuplicateColumns

File: tools/zetasql_helper/zetasql_helper/scanner/locate_table.cc [hunk 10]
```diff
@@ -31,8 +31,8 @@ absl::Status LocateTableRanges(absl::string_view query,
   ZETASQL_RETURN_IF_ERROR(ParseStatement(query, options.GetParserOptions(), &parser_output));
 
   // Predicate to find a table whose name meets the table_rex
-  auto find_table = [table_regex](const zetasql::ASTNode *node) {
-    auto table_path = dynamic_cast<const zetasql::ASTTablePathExpression *>(node);
+  auto find_table = [table_regex](const zetasql::ASTNode* node) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/locate_table.cc:34]: you can pass by reference :)

File: tools/zetasql_helper/zetasql_helper/local_service/local_service_grpc.cc [hunk 11]
```diff
@@ -90,4 +90,18 @@ grpc::Status ZetaSqlHelperLocalServiceGrpcImpl::GetAllKeywords(grpc::ServerConte
   return ToGrpcStatus(service_.GetAllKeywords(*request, response));
 }
 
+grpc::Status ZetaSqlHelperLocalServiceGrpcImpl::FixColumnNotGrouped(grpc::ServerContext* context,
+                                                                    const FixColumnNotGroupedRequest* request,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service_grpc.cc:94]: pass by reference

File: tools/zetasql_helper/zetasql_helper/fixer/fixer_test.cpp [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    EXPECT_EQ("SELECT\n"
+              "  status,\n"
+              "  max(unique_key)\n"
+              "FROM\n"
+              "  `bigquery-public-data.austin_311.311_request`\n"
+              "GROUP BY city, status\n"
+              "LIMIT 1000\n", fixed_query);
+}
+
+TEST_F(FixerTest, FixColumnNotGroupedTest3) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fixer_test.cpp:82]: in general, don't recommned using Test3, we should give it meaning full test case name
FixColumnNotGroupedWith....

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.cc [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+  auto identifier = const_cast<zetasql::ASTIdentifier*>(column_node->alias()->identifier());
+  identifier->SetIdentifier(id_string_pool->Make(new_alias));
+  ((zetasql::ASTNode*) identifier)->InitFields();
+}
+
+zetasql::ASTAlias* create_alias_node(
+    absl::string_view new_alias,
+    zetasql_base::UnsafeArena* arena,
+    zetasql::IdStringPool* id_string_pool
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.cc:147]: are we expecting to change arena and id_string_pool in the function body? if not mark const and pass by reference

File: tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.h [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+//
+// The function will assign the fixed query to the fixed_query pointer
+absl::Status
+FixColumnNotGrouped(absl::string_view query, absl::string_view missing_column, int line_number, int column_number,
+                    std::string* fixed_query);
+
+// Find the AST_Select Node having a column node. The column node should start at the given offset
+// and has the given column name.
+const zetasql::ASTSelect*
+find_select_node_having_column(const zetasql::ASTStatement* statement, int column_start_offset,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.h:42]: findSelectNodeHavingColumn

File: tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+      missing_column, parser_output->arena().get(),
+      parser_output->id_string_pool().get()
+  );
+
+  *fixed_query = Unparse(parser_output->statement());
+  return absl::OkStatus();
+}
+
+const zetasql::ASTSelect*
+find_select_node_having_column(const zetasql::ASTStatement* statement, int column_start_offset,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc:60]: FindSelectNodeHavingColumn

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.cc (hunk 13 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.cc:147]: I believe we do. We use arena to allocate space and String pool to create a new string.

File: tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc (hunk 1 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/fixer/fix_column_not_grouped.cc:28]: absl::StatusOr is not released, but I use the one defined in ZetaSQL - zetasql_base:StatusOr<T>

File: tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h (hunk 3 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/fixer/fix_duplicate_columns.h:39]: Have changed the style

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc (hunk 7 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:98]: Done

File: tools/zetasql_helper/zetasql_helper/local_service/local_service_test.cc (hunk 8 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service_test.cc:92]: Have changed all of the tests

File: tools/zetasql_helper/zetasql_helper/scanner/locate_table.cc (hunk 10 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/locate_table.cc:34]: pointer is safer here. This is predicator, so even giving a null pointer will be considered as `false`.

Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/locate_table.cc:34]: During the find process, it is impossible that a child can be null pointer.

File: tools/zetasql_helper/zetasql_helper/fixer/fixer_test.cpp (hunk 12 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/fixer/fixer_test.cpp:82]: Agree, I have changed their names

**Review by kikkyo — APPROVED** [review]

**Pull request merged**